core/IssueQueue.cpp

```diff
@@ -0,0 +1,228 @@
+}
+
+void IssueQueue::setupIssueQueue_() {
+  std::vector<core_types::RegFile> reg_files = {core_types::RF_INTEGER,
+                                                core_types::RF_FLOAT};
+  // if we ever move to multicore, we only want to have resources look for scoreboard in their cpu
+  // if we're running a test where we only have top.rename or top.issue_queue, then we can just use the root
+  auto cpu_node = getContainer()->findAncestorByName("core.*");
+  if(cpu_node == nullptr){
+    cpu_node = getContainer()->getRoot();
+  }
+  for (const auto rf : reg_files) {
+    scoreboard_views_[rf].reset(new sparta::ScoreboardView(
+        getContainer()->getName(), core_types::regfile_names[rf],
+        cpu_node));
+  }
+  out_scheduler_credits_.send(scheduler_size_);
+}
+
+void IssueQueue::onROBTerminate_(const bool &val) {
+  rob_stopped_simulation_ = val;
+}
+
+void IssueQueue::dumpDebugContent_(std::ostream &output) const {
+  output << "IssueQueue Contents" << std::endl;
+  for (const auto &entry : ready_queue_) {
+    output << '\t' << entry << std::endl;
+  }
+}
+
+void IssueQueue::onStartingTeardown_() {
+  if ((false == rob_stopped_simulation_) && (false == ready_queue_.empty())) {
+    dumpDebugContent_(std::cerr);
+    sparta_assert(false, "Issue queue has pending instructions");
+  }
+}
+
+void IssueQueue::setExePipe(std::string exe_pipe_name,
+                            olympia::ExecutePipe *exe_pipe) {
+  exe_pipes_[exe_pipe_name] = exe_pipe;
+}
+
+void IssueQueue::setExePipeMapping(InstArchInfo::TargetPipe tgt_pipe,
+                                   olympia::ExecutePipe *exe_pipe) {
+  if (pipe_exe_pipe_mapping_.find(tgt_pipe) == pipe_exe_pipe_mapping_.end()) {
+    pipe_exe_pipe_mapping_[tgt_pipe] = std::vector<olympia::ExecutePipe *>();
+  }
+  auto pipe_vector = pipe_exe_pipe_mapping_.find(tgt_pipe);
+  pipe_vector->second.emplace_back(exe_pipe);
+  pipes_.push_back(exe_pipe);
+}
+
+std::unordered_map<std::string, olympia::ExecutePipe *>
+IssueQueue::getExePipes() {
+  return exe_pipes_;
+}
+
+void IssueQueue::getInstsFromDispatch_(const InstPtr &ex_inst) {
+  sparta_assert(ex_inst->getStatus() == Inst::Status::DISPATCHED,
+                "Bad instruction status: " << ex_inst);
+  appendIssueQueue_(ex_inst);
+  handleOperandIssueCheck_(ex_inst);
+}
+
+void IssueQueue::handleOperandIssueCheck_(const InstPtr &ex_inst) {
+  // FIXME: Now every source operand should be ready
+  const auto &src_bits = ex_inst->getSrcRegisterBitMask(reg_file_);
+  if (scoreboard_views_[reg_file_]->isSet(src_bits)) {
+    // Insert at the end if we are doing in order issue or if the scheduler is
+    // empty
+    ILOG("Sending to issue queue " << ex_inst);
+    if (in_order_issue_ == true || ready_queue_.size() == 0) {
+      ready_queue_.emplace_back(ex_inst);
+    } else {
+      // Stick the instructions in a random position in the ready queue
+      uint64_t issue_pos = uint64_t(std::rand()) % ready_queue_.size();
```

Comment:
Eeek!  If someone, somewhere seeds the random generator, bugs here will be hard to reproduce.  Instead of random, how about two modes:
1. In order as they become ready (first in, first out)
2. Ordered based on age

Reply:
Implemented in [6b0e363](https://github.com/riscv-software-src/riscv-perf-model/pull/140/commits/6b0e3633c1b7331170afda8920f596f9fd16b233). Added a custom sorter to handle in fifo vs age based.